File: auth/src/main/java/com/firebase/ui/auth/util/data/EmailLinkParser.java
package com.firebase.ui.auth.util.data;

import android.net.Uri;
import android.text.TextUtils;

import com.google.android.gms.common.internal.Preconditions;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;

import androidx.annotation.NonNull;
import androidx.annotation.RestrictTo;

@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
public class EmailLinkParser {

    private static String LINK = "link";

    private static final String OOB_CODE = "oobCode";
    private static final String CONTINUE_URL = "continueUrl";

    private Map<String, String> mParams;

    public EmailLinkParser(@NonNull String link) {
        Preconditions.checkNotEmpty(link);
        mParams = parseUri(Uri.parse(link));
        if (mParams.isEmpty()) {
            throw new IllegalArgumentException("Invalid link: no parameters found");
        }
    }

    public String getOobCode() {
        return mParams.get(OOB_CODE);
    }

    public String getSessionId() {
        return mParams.get(LinkParameters.SESSION_IDENTIFIER);
    }

    public String getAnonymousUserId() {
        return mParams.get(LinkParameters.ANONYMOUS_USER_ID_IDENTIFIER);
    }

    public boolean getForceSameDeviceBit() {
        String forceSameDeviceBit = mParams.get(LinkParameters.FORCE_SAME_DEVICE_IDENTIFIER);
        if (TextUtils.isEmpty(forceSameDeviceBit)) {
            // Default value is false when no bit is set
            return false;
        }
        return forceSameDeviceBit.equals("1");
    }

    public String getProviderId() {
        return mParams.get(LinkParameters.PROVIDER_ID_IDENTIFIER);
    }

    private Map<String, String> parseUri(Uri uri) {
        Map<String, String> map = new HashMap<>();
        try {
            Set<String> queryParameters = uri.getQueryParameterNames();
            for (String param : queryParameters) {
                if (param.equalsIgnoreCase(LINK) || param.equalsIgnoreCase(CONTINUE_URL)) {
                    Uri innerUri = Uri.parse(uri.getQueryParameter(param));
                    Map<String, String> innerValues = parseUri(innerUri);
                    if (innerValues != null) {
                        map.putAll(innerValues);
                    }
                } else {
                    String value = uri.getQueryParameter(param);
                    if (value != null) {
                        map.put(param, value);
                    }
                }
            }
        } catch (Exception e) {
            // Do nothing.
        }
        return map;
    }

    public static class LinkParameters {
        public static final String SESSION_IDENTIFIER = "ui_sid";
        public static final String ANONYMOUS_USER_ID_IDENTIFIER = "ui_auid";
        public static final String FORCE_SAME_DEVICE_IDENTIFIER = "ui_sd";
        public static final String PROVIDER_ID_IDENTIFIER = "ui_pid";
    }
}

File: auth/src/main/java/com/firebase/ui/auth/util/EmailLinkParser.kt
/*
 * Copyright 2025 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the
 * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.firebase.ui.auth.util

import android.net.Uri
import androidx.annotation.RestrictTo
import androidx.core.net.toUri

/**
 * Parser for email link sign-in URLs.
 * Extracts session information and parameters from Firebase email authentication links.
 */
@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
class EmailLinkParser(link: String) {

    private val params: Map<String, String>

    init {
        require(link.isNotBlank()) { "Link cannot be empty" }
        params = parseUri(link.toUri())
        require(params.isNotEmpty()) { "Invalid link: no parameters found" }
    }

    /**
     * The out-of-band code (OOB code) from the email link.
     * This is a required field for email link authentication.
     * @throws IllegalArgumentException if the OOB code is missing from the link
     */
    val oobCode: String
        get() = requireNotNull(params[OOB_CODE]) {
            "Invalid email link: missing required OOB code"
        }

    val sessionId: String?
        get() = params[LinkParameters.SESSION_IDENTIFIER]

    val anonymousUserId: String?
        get() = params[LinkParameters.ANONYMOUS_USER_ID_IDENTIFIER]

    val forceSameDeviceBit: Boolean
        get() {
            val bit = params[LinkParameters.FORCE_SAME_DEVICE_IDENTIFIER]
            // Default value is false when no bit is set
            return bit == "1"
        }

    val providerId: String?
        get() = params[LinkParameters.PROVIDER_ID_IDENTIFIER]

    private fun parseUri(uri: Uri): Map<String, String> {
        val map = mutableMapOf<String, String>()
        try {
            val queryParameters = uri.queryParameterNames
            for (param in queryParameters) {
                if (param.equals(LINK, ignoreCase = true) ||
                    param.equals(CONTINUE_URL, ignoreCase = true)) {
                    val innerUriString = uri.getQueryParameter(param)
                    if (innerUriString != null) {
                        val innerUri = innerUriString.toUri()
                        val innerValues = parseUri(innerUri)
                        map.putAll(innerValues)
                    }
                } else {
                    val value = uri.getQueryParameter(param)
                    if (value != null) {
                        map[param] = value
                    }
                }
            }
        } catch (e: Exception) {
            // Do nothing - return what we have
        }
        return map
    }

    object LinkParameters {
        const val SESSION_IDENTIFIER = "ui_sid"
        const val ANONYMOUS_USER_ID_IDENTIFIER = "ui_auid"
        const val FORCE_SAME_DEVICE_IDENTIFIER = "ui_sd"
        const val PROVIDER_ID_IDENTIFIER = "ui_pid"
    }

    companion object {
        private const val LINK = "link"
        private const val OOB_CODE = "oobCode"
        private const val CONTINUE_URL = "continueUrl"
    }
}